fl4health/clients/deep_mmd_clients/ditto_deep_mmd_client.py

```diff
@@ -0,0 +1,201 @@
+        self.init_global_feature_extractor._maybe_register_hooks()
+
+    def predict(
+        self,
+        input: TorchInputType,
+    ) -> Tuple[Dict[str, torch.Tensor], Dict[str, torch.Tensor]]:
+        """
+        Computes the predictions for both the GLOBAL and LOCAL models and pack them into the prediction dictionary
+
+        Args:
+            input (Union[torch.Tensor, Dict[str, torch.Tensor]]): Inputs to be fed into both models.
+
+        Returns:
+            Tuple[Dict[str, torch.Tensor], Dict[str, torch.Tensor]]: A tuple in which the first element
+            contains predictions indexed by name and the second element contains intermediate activations
+            index by name.
+
+        Raises:
+            ValueError: Occurs when something other than a tensor or dict of tensors is returned by the model
+            forward.
+        """
+
+        # We use features from init_global_model to compute the DEEP-MMD loss not the global_model
+        global_preds = self.global_model(input)
+        local_preds = self.model(input)
+        features = self.local_feature_extractor.get_extracted_features()
+        if self.deep_mmd_loss_weight != 0:
+            # Compute the features of the init_global_model
+            _ = self.init_global_model(input)
+            init_global_features = self.init_global_feature_extractor.get_extracted_features()
+            for key in init_global_features.keys():
+                features[" ".join(["init_global", key])] = init_global_features[key]
+
+        return {"global": global_preds, "local": local_preds}, features
+
+    def _maybe_checkpoint(self, loss: float, metrics: Dict[str, Scalar], checkpoint_mode: CheckpointMode) -> None:
+        # Hooks need to be removed before checkpointing the model
+        self.local_feature_extractor.remove_hooks()
+        super()._maybe_checkpoint(loss=loss, metrics=metrics, checkpoint_mode=checkpoint_mode)
+
+    def compute_evaluation_loss(
+        self,
+        preds: Dict[str, torch.Tensor],
+        features: Dict[str, torch.Tensor],
+        target: torch.Tensor,
+    ) -> EvaluationLosses:
+        """
+        Computes evaluation loss given predictions (and potentially features) of the model and ground truth data.
+        We also need to set the training flag of the deep_mmd_losses to False to ensure that the model is in eval mode.
+
+        Args:
+            preds (Dict[str, torch.Tensor]): Prediction(s) of the model(s) indexed by name. Anything stored
+                in preds will be used to compute metrics.
+            features: (Dict[str, torch.Tensor]): Feature(s) of the model(s) indexed by name.
+            target: (torch.Tensor): Ground truth data to evaluate predictions against.
+
+        Returns:
+            EvaluationLosses: an instance of EvaluationLosses containing checkpoint loss and additional losses
+                indexed by name.
+        """
+        for layer in self.flatten_feature_extraction_layers.keys():
```

Comment:
If you're going to be indexing into `self.deep_mmd_losses` anyway, could we simply do 
```
for layer_loss_module in self.deep_mmd_losses.values():
    layer_loss_module.training = False
```
Correct me if I'm wrong, but you likely also want to do something like `layer_loss_module.requires_grad=False as well here?

Reply:
I think it's still worth overriding `compute_evaluation_loss` and `compute_training_loss` and asserting that all `layer_loss_module.training == False` or vice versa though to be safe 🙂

Reply:
I also might be missing this, but I don't see where we set `layer_loss_module.training` to `True` in the client. Based on the loss code, this would mean that we won't run training of the deep kernels after the first server round, which I think we want to keep doing?

Reply:
Good catch! The `True` setting was indeed missing, so I added it to the `update_before_train` function. Following your suggestion, I moved the `False` setting to the `validate` function. I kept the assertions in both `compute_evaluation_loss` and `compute_training_loss` functions for consistency.

Reply:
Soo now after doing that, the optimization of kernel parameters and model will happen in one step but with different optimizers, however there is a potential second approach can be to separate these optimizations. Meaning that during training for each batch first we freeze model and optimize kernel, then freeze kernel and optimize model.